Implement constant support in MIR.

eddyb opened the pull request:
All of the intended features in `trans::consts` are now supported by `mir::constant`.
The implementation is considered a temporary measure until `miri` replaces it.

A `-Z orbit` bootstrap build will only translate LLVM IR from AST for `#[rustc_no_mir]` functions.

Furthermore, almost all checks of constant expressions have been moved to MIR.
In non-`const` functions, trees of temporaries are promoted, as per RFC 1414 (rvalue promotion).
Promotion before MIR borrowck would allow reasoning about promoted values' lifetimes.

The improved checking comes at the cost of four `[breaking-change]`s:
* repeat counts must contain a constant expression, e.g.:
`let arr = [0; { println!("foo"); 5 }];` used to be allowed (it behaved like `let arr = [0; 5];`)
* dereference of a reference to a `static` cannot be used in another `static`, e.g.:  
`static X: [u8; 1] = [1]; static Y: u8 = (&X)[0];` was unintentionally allowed before
* the type of a `static` *must* be `Sync`, irrespective of the initializer, e.g.  
`static FOO: *const T = &BAR;` worked as `&T` is `Sync`, but it shouldn't because `*const T` isn't
* a `static` cannot wrap `UnsafeCell` around a type that *may* need drop, e.g.  
`static X: MakeSync<UnsafeCell<Option<String>>> = MakeSync(UnsafeCell::new(None));`  
was previously allowed based on the fact `None` alone doesn't need drop, but in `UnsafeCell`
it can be later changed to `Some(String)` which *does* need dropping

The drop restrictions are relaxed by RFC 1440 (#33156), which is implemented, but feature-gated.
However, creating `UnsafeCell` from constants is unstable, so users can just enable the feature gate.

rust-highfive commented:
r? @arielb1

(rust_highfive has picked a reviewer for you, use r? to override)

oli-obk commented:
I don't understand the reason for the existance of ce084d5

Won't all of that code go away once miri does the interpretation?

eddyb commented:
Yes it would, I thought it was going to be less code. I just wanted to make sure that we can work independently of old trans, no matter what happens with miri integration.

arielb1 commented on src/librustc_mir/mir_map.rs:
```diff
@@ -58,89 +56,63 @@ struct BuildMir<'a, 'tcx: 'a> {
                 id: ast::NodeId) {
-        let implicit_arg_tys = if let intravisit::FnKind::Closure(..) = fk {
-            vec![closure_self_ty(&self.tcx, id, body.id)]
+        // fetch the fully liberated fn signature (that is, all bound
```
Shouldn't you `.subst(free_subst)`?

eddyb replied:
It's the same code as before, just moved around.

arielb1 replied:
that's the Correct way

nikomatsakis commented:
cc @rust-lang/compiler

nikomatsakis commented:
I'd like to sign off one this before it is r+'d. Will try to look today.

nagisa commented on src/test/compile-fail/feature-gate-negate-unsigned.rs:
```diff
@@ -16,7 +16,7 @@ impl std::ops::Neg for S {
 }
 
-const _MAX: usize = -1;
+const _MAX: usize = -(2 - 1);
```
??

eddyb replied:
MIR errors while building while the test expects an error emitted during lint-checking (but not suppressible AFAICT), and the driver stops on errors somewhere in between.

nagisa replied:
Should add a `FIXME` comment to return to the original code once it is possible (i.e. the current consteval code is removed).

nagisa commented on src/test/run-pass/const-autoderef.rs:
```diff
@@ -9,10 +9,10 @@
+const A: [u8; 1] = ['h' as u8];
+const B: u8 = (&A)[0];
+const C: &'static &'static &'static &'static [u8; 1] = & & & &A;
+const D: u8 = (&C)[0];
```
Why change test from ages ago? Are we even sure that despite the test name this test was not testing precisely for behaviour of `static`?

eddyb replied:
This behavior was never intentionally supported AFAIK, just the test didn't get changed from `static` to `const`. cc @oli-obk

oli-obk replied:
It should be forbidden, but crates might be depending on it. We probably need a warning cycle for this kind of breaking change

eddyb replied:
I'll keep that in mind (and get a crater run) for my next PR which I hope to move all constant checking to MIR.

nagisa commented on src/librustc_trans/mir/block.rs:
```diff
@@ -114,9 +114,16 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
+                        ty: switch_ty,
+                        span: terminator.span
+                    };
+                    let val = self.trans_constant(&bcx, &constant).immediate();
```
why not use the `Const::from_constval`? Seems weird to wrap the constval back into some structure only to have it immediately unwrapped later.

eddyb replied:
I wanted to avoid making `Const` public as it's part of the code that would go away with miri (and there would be a different way to turn a miri primitive value into a LLVM constant).

nagisa commented on src/librustc_trans/consts.rs:
```diff
@@ -1147,14 +1142,20 @@ pub fn trans_static(ccx: &CrateContext,
+            let default_to_mir = ccx.sess().opts.debugging_opts.orbit;
+            let invert = if default_to_mir { "rustc_no_mir" } else { "rustc_mir" };
+            default_to_mir ^ attrs.iter().any(|item| item.check_name(invert))
+        };
```
I feel like this is duplicating the logic used for functions and is very likely to diverge (developers are likely to check their changes work for MIR functions and unlikely to check they work for MIR constants). Perhaps factor it out into a simple function in `trans::common`?

eddyb replied:
I believe @michaelwoerister's work on driving trans from the collector will supersede these checks, or at least I hope so.

bors commented:
:umbrella: The latest upstream changes (presumably #33179) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
Crater run: https://gist.github.com/nikomatsakis/c2c7ecdac8437dc8056b6629ec7e9f83

* 4063 crates tested: 2631 working / 1322 broken / **3 regressed / 0 fixed** / 107 unknown.

eddyb commented:
@nikomatsakis one is a timeout, one is a non-`Sync` `static` (I... didn't do anything intentional there, how could that have escaped checking before?) and one (`odbc`) is a true regression:
```rust
src/raw/sqlconsts.rs:272:48: 272:71 error: cannot refer to other statics by value, use the address-of operator or a cons
tant instead [E0394]                                                                                                    
src/raw/sqlconsts.rs:272 pub static SQL_MAXIMUM_USER_NAME_LENGTH: u16 = *&SQL_MAX_USER_NAME_LEN;                        
                                                                        ^~~~~~~~~~~~~~~~~~~~~~~ 
```
Followed by 22 more of these, which appear to be *aliases* (`use` could've worked!) of other `static`s which should also be `const` instead.
I consider this to be pathological abuse of a lack of rigor in our checking and I don't think there's a way to have a warning cycle - besides, it can't compile with MIR trans.

I will personally try to fix these two crates, I hope we can merge this soon, since it's blocking both miri and some of my other `'tcx`-related work.

eddyb commented:
Actually, the misuse of `static` from `obdc` was easy to fix, see Koka/odbc-rs#1 (already merged).

However, the non-`Sync` `static` from `dispatch` is worse, because it *is* a regression, one I have not considered:
```rust
static FOO: *const i32 = &0;
```
We allow this only because we look at the *unadjusted* initializer type, which is `&'static T`.
This type isn't even available in the MIR, so I don't see a point in trying to keep this working, but I'll add it to the list in the PR description.

bors commented:
:umbrella: The latest upstream changes (presumably #32791) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented on src/librustc_mir/transform/promote_consts.rs:
```diff
@@ -0,0 +1,404 @@
+                        };
+                        return;
+                    }
+                    _ => {}
```
Should this a `bug` case?

nikomatsakis replied:
(I suspect it should, actually, since we are visiting in RPO)

eddyb replied:
I agree, promotion can make invalid code pass borrowck otherwise (as per the comment at the top).

eddyb replied:
Actually, this can't really work because of how `LvalueContext` behaves around projections.

nikomatsakis commented on src/librustc_mir/transform/promote_consts.rs:
```diff
@@ -0,0 +1,404 @@
+                        *uses += 1;
+                        return;
+                    }
+                    _ => {}
```
Do we want to track when the temp is multiply assigned?

eddyb replied:
AFAIK multiple temp assignment right now is only due to conditional value-flow, i.e. `if`-`else` or `match`, which promotion doesn't support atm. If we do want to do it in the future, we'll have to use a proper dataflow analysis framework for it.

nikomatsakis commented on src/librustc_mir/transform/promote_consts.rs:
```diff
@@ -0,0 +1,404 @@
+                    _ => {}
+                }
+            }
+            *temp = TempState::Unpromotable;
```
Ah, perhaps disregard my prior two comments -- but add to the `{}` cases above something like:

```
{ /* mark as unpromotable below */ }
```

eddyb commented on src/librustc_mir/transform/qualify_consts.rs:
```diff
@@ -222,13 +222,36 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx> {
+        }
+
+        // Feature-gate for globals with destructors is enabled.
+        if self.tcx.sess.features.borrow().drop_types_in_const {
```
Note to self: don't actually allow this in `const` items (as per RFC).

nikomatsakis commented on src/librustc_mir/transform/qualify_consts.rs:
```diff
@@ -0,0 +1,984 @@
+                }
+
+                TerminatorKind::If { targets: (if_true, if_else), .. } => {
+                    // Need to allow bounds checking branches.
```
Eurgh. This suggests to me that we may want to have special terminators for bounds-checks. (We could lower them to calls later on.)

I feel like this might make sense for other reasons -- e.g., the current MIR builder code code builds a lot of random tuples that seem expensive to build for normal cases, iirc.

nikomatsakis replied:
(e.g. note that "hand-unrolled" bounds checks will pass this check too, which seems not necessarily right.)

eddyb replied:
At some point I mentioned a complex terminator scheme which multiplexes conditions and choices (i.e. you could have a `Switch` with a `Call`, `Drop` or `Br` for each variant).
In any case, I'd be down with a special terminator if it's used for arithmetic checks as well.
cc @Aatch

nikomatsakis commented on src/librustc_mir/transform/qualify_consts.rs:
```diff
@@ -0,0 +1,984 @@
+                            }), ..
+                        } => {
+                            if Some(def_id) == tcx.lang_items.panic_bounds_check_fn() {
+                                Some(if_true)
```
We should at least check that the list of statements in the `if_else` branch is empty.

eddyb replied:
That's not actually the case, there are at least one or two statements involved to create the arguments.

nikomatsakis commented on src/librustc_mir/transform/qualify_consts.rs:
```diff
@@ -0,0 +1,984 @@
+    // FIXME(eddyb) we could split the errors into meaningful
+    // categories, but enabling full miri would make that
+    // slightly pointless (even with feature-gating).
+    fn unimplemented(&mut self) {
```
can you rename this to `not_const`? I was confused reading this code for some time because I assumed that `unimplemented` would be equivalent to `bug!`, and it seemed like this would cause a fair number of panics.

nikomatsakis commented:
Sorry for being slow. I've not read this in great detail, but enough to get a sense of the overall approach. It seems good to me! :) r+ from me modulo nits. (I think there was another r+ already anyhow...)

bors commented:
:umbrella: The latest upstream changes (presumably #33303) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented:
@nikomatsakis I've added code to create `DefId`s and build MIR for constant integers in `enum` discriminants, fixed-size array types and array repeat expressions, the last of which wasn't checked at all before.
Mind taking a look at it?

bors commented:
:umbrella: The latest upstream changes (presumably #32756) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
@eddyb is that in some specific commit, or sprinkled throughout?

nikomatsakis commented:
r @nikomatsakis

nikomatsakis commented:
@eddyb r=me

eddyb commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 12f8a1d has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 12f8a1d4f9fc987bc29b2a3f68c3745c8d7d0fd8 -->

bors commented:
:hourglass: Testing commit 12f8a1d with merge 7134375...

bors commented:
:broken_heart: Test failed - [auto-linux-64-cross-freebsd](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/192)

bors commented:
:umbrella: The latest upstream changes (presumably #33138) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit da967f0 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis da967f04839fd35accdfcc5aaea0d1a1f57c1ef4 -->

eddyb commented:
@bors r-

bors commented:
:pushpin: Commit da967f0 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis da967f04839fd35accdfcc5aaea0d1a1f57c1ef4 -->

eddyb commented:
@bors r-

eddyb commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit beeb429 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis beeb429f1529bf5bbfded2796f376d47c02990c8 -->

bors commented:
:lock: Merge conflict

bors commented:
:umbrella: The latest upstream changes (presumably #33228) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit b765580 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis b7655804062f63dcc20784c13f1c3a9e2c981b36 -->

bors commented:
:umbrella: The latest upstream changes (presumably #33267) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 29dd7e0 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 29dd7e02a8e812ee54b58d8cec633fd53ddb4007 -->

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #32900

bors commented:
:pushpin: Commit 29dd7e0 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 29dd7e02a8e812ee54b58d8cec633fd53ddb4007 -->

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #32900

bors commented:
:pushpin: Commit 29dd7e0 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 29dd7e02a8e812ee54b58d8cec633fd53ddb4007 -->

eddyb commented:
@bors r-

bors commented:
:pushpin: Commit 29dd7e0 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 29dd7e02a8e812ee54b58d8cec633fd53ddb4007 -->

eddyb commented:
@bors r-

eddyb commented:
@bors r=nikomatsakis

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #32900

bors commented:
:pushpin: Commit 29dd7e0 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 29dd7e02a8e812ee54b58d8cec633fd53ddb4007 -->

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #32900

bors commented:
:pushpin: Commit 29dd7e0 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 29dd7e02a8e812ee54b58d8cec633fd53ddb4007 -->

eddyb commented:
@bors r-

eddyb commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 29dd7e0 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 29dd7e02a8e812ee54b58d8cec633fd53ddb4007 -->

bors commented:
:hourglass: Testing commit 29dd7e0 with merge ebf4761...

bors commented:
:broken_heart: Test failed - [auto-linux-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/1008)

arielb1 commented:
@bors retry

bors commented:
:hourglass: Testing commit 29dd7e0 with merge 41f3d25...

bors commented:
:broken_heart: Test failed - [auto-linux-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/1012)

eddyb commented:
Ahh, it's a real error caused by the changes to how MIR passes are ran.

eddyb commented:
@bors retry

eddyb commented:
@bors force

bors commented:
:zap: Previous build results for [auto-win-msvc-64-cargotest](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/396) are reusable. Rebuilding only [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/9080), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/9117), [auto-linux-32cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-32cross-opt/builds/200), [auto-linux-64-cargotest](http://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/392), [auto-linux-64-cross-armhf](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-armhf/builds/203), [auto-linux-64-cross-armsf](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-armsf/builds/207), [auto-linux-64-cross-freebsd](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/206), [auto-linux-64-cross-netbsd](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-netbsd/builds/206), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/2349), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/9055), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/9041), [auto-linux-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-mir/builds/627), [auto-linux-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/1012), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/9037), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/2248), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/4192), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/9093), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/9102), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/9063), [auto-mac-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/1021), [auto-mac-cross-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/205), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/4128), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/4141), [auto-win-gnu-32-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/1025), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/4134), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/4117), [auto-win-msvc-32-cross-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-cross-opt/builds/489), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3338), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/4130), [auto-win-msvc-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-mir/builds/630), [auto-win-msvc-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/1028)...

bors commented:
:broken_heart: Test failed - [auto-win-gnu-32-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/1027)

eddyb commented:
@alexcrichton Is this spurious?
I thought it might be my fault but I looked at the implementation and short of RBML corruption, I don't see how these failures could happen.

eddyb commented:
cc @michaelwoerister

arielb1 commented:
Do the failures happen locally?

eddyb commented:
@arielb1 Nope, although I don't use rustbuild which may be relevant.

eddyb commented:
@bors retry

The tests passed under rustbuild locally.

eddyb commented:
@bors r=nikomatsakis

Updated `Cargo.lock` (not that it should be relevant to the failures).

bors commented:
:pushpin: Commit 5c42e69 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 5c42e694cb7eb2bc26336351a1fb4e1e4f94768e -->

bors commented:
:hourglass: Testing commit 5c42e69 with merge 45897b0...

alexcrichton commented:
@eddyb 

* [This] was spurious - https://github.com/rust-lang/rust/issues/33435

Oh and looks like the other two were the same

bors commented:
:broken_heart: Test failed - [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/9107)

eddyb commented:
[Actual failure](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/1017/steps/test/logs/stdio), same one.

arielb1 commented:
Does this not reproduce locally?

eddyb commented:
@arielb1 Neither me nor Travis CI can reproduce it on linux.

eddyb commented:
Choosing the nuclear option to enable the rustc dev sprint next week.

@bors r=nikomatsakis p=99

bors commented:
:pushpin: Commit 97e8e7f has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 97e8e7f1510c37cf2f300da511bc379f64f873b9 -->

eddyb commented:
Or not, maybe we can get away with simply removing the directory before running the tests.

@bors r=nikomatsakis

bors commented:
:pushpin: Commit 3001626 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 300162659504fed4798e535791023bf0174fc81c -->

bors commented:
:hourglass: Testing commit 3001626 with merge 6c7f924...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/1039)

eddyb commented:
[Actual failure](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-mir/builds/637/steps/test/logs/stdio).

eddyb commented:
@bors r=nikomatsakis

@alexcrichton Remind me to bug you about enabling LLVM assertions on the test bots, something that should've been caught by an assertion ended up in a successful compile (but the resulting program crashed in an edge condition).
This seems very dangerous to have around, as we keep fiddling with trans.

bors commented:
:pushpin: Commit 3b0e27c has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 3b0e27cc74d61e229aeaf0a710d3a018f7104ffc -->

bors commented:
:hourglass: Testing commit 3b0e27c with merge 1ec2217...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/9098), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/9135), [auto-linux-32cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-32cross-opt/builds/218), [auto-linux-64-cargotest](http://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/410), [auto-linux-64-cross-armhf](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-armhf/builds/221), [auto-linux-64-cross-armsf](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-armsf/builds/225), [auto-linux-64-cross-freebsd](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/224), [auto-linux-64-cross-netbsd](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-netbsd/builds/224), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/2367), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/9073), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/9059), [auto-linux-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-mir/builds/645), [auto-linux-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/1030), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/9055), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/2266), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/4210), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/9111), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/9120), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/9081), [auto-mac-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/1039), [auto-mac-cross-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/223), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/4146), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/4159), [auto-win-gnu-32-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/1043), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/4153), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/4135), [auto-win-msvc-32-cross-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-cross-opt/builds/508), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3356), [auto-win-msvc-64-cargotest](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/413), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/4148), [auto-win-msvc-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-mir/builds/648), [auto-win-msvc-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/1047)

alexcrichton commented:
@eddyb oh yeah I'd be fine enabling LLVM assertions, we'd just want to time it and ensure it doesn't regress cycle time too much. If it's just a few minutes that's fine though as it's some good testing. 

I thought we already did this but apparently not! Looks like we only enable LLVM assertions on nightly.

Ryman commented:
The following seems to have regressed, but I'm unsure if that's intentional or not? (also not entirely sure if this is the correct PR but it looks most related)  

```rust
pub static mut freelist: [&'static mut [u8]; 16] =
    [&mut [], &mut [], &mut [], &mut [], &mut [], &mut [], &mut [],
    &mut [], &mut [], &mut [], &mut [], &mut [], &mut [], &mut [],
    &mut [], &mut []];
```

Runs on stable and beta, fails on nightly: https://is.gd/9tUBN8

This is *in the wild* but not on crates.io, https://github.com/dropbox/rust-alloc-no-stdlib/issues/1 which is a dependency for https://github.com/dropbox/rust-brotli-no-stdlib

eddyb commented:
@Ryman That is supposed to be allowed, but I can't see how it happens, because it's supposed to be whitelisted based on the type.

eddyb commented:
@Ryman Should be fixed in #33620.
